baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/FinalClass.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * (c) Copyright 2019 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.baseline.errorprone;
+
+import com.google.auto.service.AutoService;
+import com.google.errorprone.BugPattern;
+import com.google.errorprone.VisitorState;
+import com.google.errorprone.bugpatterns.BugChecker;
+import com.google.errorprone.fixes.SuggestedFixes;
+import com.google.errorprone.matchers.Description;
+import com.google.errorprone.util.ASTHelpers;
+import com.sun.source.tree.ClassTree;
+import com.sun.source.tree.MethodTree;
+import com.sun.source.tree.NewClassTree;
+import com.sun.source.tree.Tree;
+import com.sun.source.util.TreeScanner;
+import java.util.List;
+import java.util.Set;
+import javax.lang.model.element.Modifier;
+
+@AutoService(BugChecker.class)
+@BugPattern(
+        name = "FinalClass",
+        // Support legacy suppressions from checkstyle
+        altNames = {"checkstyle:finalclass", "checkstyle:FinalClass"},
+        link = "https://github.com/palantir/gradle-baseline#baseline-error-prone-checks",
+        linkType = BugPattern.LinkType.CUSTOM,
+        providesFix = BugPattern.ProvidesFix.REQUIRES_HUMAN_ATTENTION,
+        severity = BugPattern.SeverityLevel.ERROR,
+        summary = "A class should be declared final if all of its constructors are private. Utility classes -- "
+                + "i.e., classes all of whose methods and fields are static -- have a private, empty, "
+                + "zero-argument constructor.\n"
+                + "https://github.com/palantir/gradle-baseline/tree/develop/docs/best-practices/"
+                + "java-coding-guidelines#private-constructors")
+public final class FinalClass extends BugChecker implements BugChecker.ClassTreeMatcher {
```

Comment:
we may want to consider checking/cleaning up all of the methods in the class to avoid inadvertently causing method modifiers to become redundant

Reply:
I like that, remove final from methods that inherit finality from the class. Error prone doesn't have an unnecessary modifier check afaik, thoughts on implementing that separately? Counterargument is that it must be applied in another round after this check succeeds, but I like the separation of validation, and the ability to fix existing instances that aren't created by this check.

Reply:
It would be nice to have both. A generic redundant modifier check and then a special cleanup on this check to specifically remove `final` from methods of classes that have been converted. That way we avoid having to run the fix twice to get to a good state and can catch other issues like `public` on interface methods

Reply:
Implemented the fix, four hits in large internal ;-)

Reply:
Bug: We can't remove the `final` modifier from methods annotated with `java.lang.SafeVarargs`